### ISSUE TITLE
Running Neovim in Vim's terminal triggers E5009 (uses Vim's VIM/VIMRUNTIME)

### ISSUE THREAD

**Issue by blueyed** [neovim/neovim#8696]
Vim appears to set `$VIM` and `$VIMRUNTIME` in its terminal environment, which then gets picked up by Neovim.

Running `:checkhealth` will trigger:

> E5009: Invalid $VIMRUNTIME: /usr/share/vim/vim81

The workaround is to unset VIM/VIMRUNTIME before running nvim.

Vim 8.1.22
NVIM v0.3.1-139-g2bfabd5b

**Comment by blueyed** [comment]
Vim behaves the same (it also uses an existing $VIM/VIMRUNTIME from Neovim's terminal), but there `has('python3')` still works.

Not sure if anything can be improved here really - feel free to close it.

**Comment by justinmk** [comment]
Yeah, this is a pain point for me. We need to undefine VIMRUNTIME, NVIM_LISTEN_ADDRESS, etc. in `:terminal`. Been waiting on https://github.com/neovim/neovim/pull/7202 but maybe we can do something else in the meantime.

**Comment by justinmk** [comment]
Oops I got your report mixed up. I thought Vim already undefined VIMRUNTIME in its `:terminal`. There's not really anything we can do about Vim's `:terminal` ....

**Comment by blueyed** [comment]
> undefine VIMRUNTIME, NVIM_LISTEN_ADDRESS, etc.

Maybe it would be better to not pick it up during startup?
It might be useful to see if there is an outer (Neo)vim after all.
Maybe VIMRUNTIME could be unset, but not VIM or something in this direction - or another new environment variable could be set to indicate that this is `:terminal`?

I came across this when trying to improve vim-tmux-navigator when used nested (via https://github.com/christoomey/vim-tmux-navigator/pull/201).  But there appears to be no autocommand for when an inner Neovim is entered it seems.  Maybe FocusGained (should) get(s) triggered?! (i.e. the outer Neovim should send the escape sequences when entering/leaving a terminal window?! (if focus reporting is enabled)

**Comment by janlazo** [comment]
@blueyed Did you try detecting `VIM_SERVERNAME`?

**Comment by aspieln3r** [comment]
I would like to work on this issue.
The solution seems to be checking for `$VIM_TERMINAL` and then if set, unset `$VIM` and `$VIMRUNTIME`. 
Can someone guide me from here?

**Comment by drosan-dev** [comment]
I unset VIM VIMRUNTIME when nvim start

**Comment by drosan-dev** [comment]
https://github.com/neovim/neovim/pull/12759

**Comment by shinymimiq** [comment]
unset `VIM` and `VIMRUNTIME` will still cause problem when call `:checkhealth` in Nvim inside of vim terminal because we need valid VIM and VIMRUNTIME for `:checkhealth`

What about we handle this special case inside our `vim_getenv` function. 
We can check if we are trying to get `VIMRUNTIME` or `VIM` inside of vim terminal, if so, we should assume those are empty and set them as we do normally. 
e.g #14838

**Comment by jamessan** [comment]
> We can check if we are trying to get `VIMRUNTIME` or `VIM` inside of vim terminal, if so, we should assume those are empty and set them as we do normally.

What if I'm trying to run against a custom runtime from with a vim terminal?  `VIMRUNTIME=... nvim foo` would be the typical way to do that, but you're suggesting to just ignore the explicitly set `VIMRUNTIME` because `VIM_TERMINAL` is set?

There are really two problems here.

* nvim stuck with the same environment variables Vim uses.
* Both programs export `VIM`/`VIMRUNTIME` to the environments of their child processes

I don't recall (and don't feel like looking right now) whether the latter point has been discussed for Vim.  However, it seems those two variables should _not_ be exported into child processes. Then we would know for sure that if they're set, it's because the user did so.

`VIM_SERVERNAME`/`NVIM_LISTEN_ADDRESS` should still be exported to child processes, since that allows communicating with the containing process (including querying them for their `VIMRUNTIME` if needed).

**Comment by shinymimiq** [comment]
> What if I'm trying to run against a custom runtime from with a vim terminal? VIMRUNTIME=... nvim foo would be the typical way to do that, but you're suggesting to just ignore the explicitly set VIMRUNTIME because VIM_TERMINAL is set?

I didn't actually considered this case. Looks like the approach discussed in this thread won't resolve this issue. 

> - nvim stuck with the same environment variables Vim uses.
> - Both programs export VIM/VIMRUNTIME to the environments of their child processes

Solutions might come to these two. 
1. Use different environment variables in neovim
2. Vim don't export VIM/VIMRUNTIME to the environments of their child processes.

**Comment by clason** [comment]
Duplicate of https://github.com/neovim/neovim/issues/6764, so I'm closing this one. (Feel free to reopen and close the other one instead.)

**Comment by zeertzjq** [comment]
No, this is not a duplicate. This is about running Nvim in Vim's terminal, that one is about running Vim in Nvim's terminal.

**Comment by clason** [comment]
Whoops! Only consolation is that Justin made the same mistake (https://github.com/neovim/neovim/issues/8696#issuecomment-403109423) 😄

**Comment by justinmk** [comment]
Nothing we can do about this. Not a bug. Could be a feature request to Vim.
